add print-manifest flag to print addon manifests to STDOUT

wangyysde opened the pull request:
Signed-off-by: wangyysde <net_use@bzhy.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
 It can be useful to allow kubeadm users to skip addons but still be able to obtain the addon manifests that kubeadm will apply for version X. This PR add print-manifest flag to support printing addon manifests to STDOUT.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes kubernetes/kubeadm#2681

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kubeadm: add support for the flag "--print-manifest" to the addon phases "kube-proxy" and "coredns" of "kubeadm init phase addon". If this flag is used kubeadm will not apply a given addon and instead print to the terminal the API objects that will be applied.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -290,6 +292,11 @@ func AddInitOtherFlags(flagSet *flag.FlagSet, initOptions *initOptions) {
 		&initOptions.skipCertificateKeyPrint, options.SkipCertificateKeyPrint, initOptions.skipCertificateKeyPrint,
 		"Don't print the key used to encrypt the control-plane certificates.",
 	)
+	flagSet.BoolVar(
+		&initOptions.printManifest, options.PrintManifest, initOptions.printManifest,
+		"Print the addon manifests to STDOUT instead of  install them.",
+	)
```
i think we should not add the flag to the main `init` or `init phase addon`, but *only* to the addon phases:
- `init phase addon coredns --print-manifest`
- `init phase addon kube-proxy --print-manifest`

wangyysde replied:
Thanks. I will move the flag to the addon phases

pacoxu replied:
`Print the addon manifests to STDOUT instead of  install them.`
->
`Print the addon manifest to STDOUT instead of installing them.`

neolit123 replied:
The plural form is correct here (with s). We can adjust the message again once the rest of the code org is done.

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -132,6 +132,15 @@ func coreDNSAddon(cfg *kubeadmapi.ClusterConfiguration, client clientset.Interfa
 	}
 
+	if printManifest {
+		fmt.Printf("%s\n", coreDNSDeploymentBytes)
+		fmt.Println("---")
+		fmt.Printf("%s\n", coreDNSConfigMapBytes)
+		fmt.Println("---")
+		fmt.Printf("%s\n", coreDNSServiceBytes)
+		return nil
```
we should start with initial `---` on the first line:

```suggestion
		fmt.Println("---")
		fmt.Printf("%s\n", coreDNSDeploymentBytes)
		fmt.Println("---")
		fmt.Printf("%s\n", coreDNSConfigMapBytes)
		fmt.Println("---")
		fmt.Printf("%s\n", coreDNSServiceBytes)
		return nil
```

neolit123 commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -118,6 +118,12 @@ func createKubeProxyConfigMap(cfg *kubeadmapi.ClusterConfiguration, localEndpoin
 	}
 
+	if printManifest {
+		fmt.Printf("%s\n", configMapBytes)
+		fmt.Println("---")
```
```suggestion
		fmt.Println("---")
		fmt.Printf("%s\n", configMapBytes)
```

we should print the deamonset here too?

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -132,6 +132,15 @@ func coreDNSAddon(cfg *kubeadmapi.ClusterConfiguration, client clientset.Interfa
 		return errors.Wrap(err, "error when parsing CoreDNS service template")
 	}
 
+	if printManifest {
```
i think this should be in EnsureDNSAddon

neolit123 commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -47,16 +47,16 @@ const (
 
 // EnsureProxyAddon creates the kube-proxy addons
-func EnsureProxyAddon(cfg *kubeadmapi.ClusterConfiguration, localEndpoint *kubeadmapi.APIEndpoint, client clientset.Interface) error {
+func EnsureProxyAddon(cfg *kubeadmapi.ClusterConfiguration, localEndpoint *kubeadmapi.APIEndpoint, client clientset.Interface, printManifest bool) error {
```
instead of using `fmt.Printf/ln` we should use pass a io.Writer to EnsureProxyAddon and EnsureDNSAddon functions.
and use 

kubeadm commands pass io.Writer from the command construction:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/kubeadm.go#L49

neolit123 reviewed (commented):
thanks for the PR @wangyysde 
i've added a few comments.

basically what we want is the following:
- add the new flag only to the addon phases them self, but not to the parent `addon` command or `init`.
- `Ensure*` addon functions can accept the `printManifest` bool flag, but do the printing on an `io.Writer` and not always to STDOUT.

that seems like the easiest refactor.
please let me know if you have questions.

/cc @pacoxu
PTAL for reviews too if you have time.

neolit123 commented:
/remove-area code-generation provider/gcp
/ kind feature
/priority backlog
/triage accepted
/remove-sig api-machinery cloud-provider

neolit123 commented:
/release-note-edit

```release-note
kubeadm: add support for the flag "--print-manifest" to the addon phases "kube-proxy" and "coredns" of "kubeadm init phase addon". If this flag is used kubeadm will not apply a given addon and instead print to the terminal the API objects that will be applied.
```

pacoxu commented:
/kind feature

pacoxu commented:
I build kubeadm and test it on my cluster, and it doesn't work.

You can try to build like `build/run.sh make kubeadm`.

```
[root@paco ~]# ./kubeadm init --print-manifest=true  phase addon coredns -v 4
unknown flag: --print-manifest
To see the stack trace of this error execute with --v=5 or higher
[root@paco ~]# ./kubeadm init  phase addon coredns --print-manifest=true
unknown flag: --print-manifest
To see the stack trace of this error execute with --v=5 or higher
```

wangyysde committed:
add print-manifest flag to print addon manifests to STDOUT

Signed-off-by: wangyysde <net_use@bzhy.com>

pacoxu commented on cmd/kubeadm/app/cmd/options/constant.go:
```diff
@@ -145,4 +145,7 @@ const (
 	Patches = "patches"
+
+	// Print the addon manifests to STDOUT instead of installing them.
+	PrintManifest = "print-manifest"
```
The behavior is quite similar to `dry-run`. So I compare the output of dry-run of `kubeadm init` with only addon phase.

```
./kubeadm init --dry-run --skip-phases=preflight,kubelet-start,control-plane,etcd,upload-certs,mark-control-plane,bootstrap-token,kubelet-finalize,wait-control-plane
```

The output is similar. However, I find that rbac resources are missing. 

```
[dryrun] Resource name: "coredns"
[dryrun] Would perform action CREATE on resource "configmaps" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	data:
	  Corefile: |
	    .:53 {
	        errors
	        health {
	           lameduck 5s
	        }
	        ready
	        kubernetes cluster.local in-addr.arpa ip6.arpa {
	           pods insecure
	           fallthrough in-addr.arpa ip6.arpa
	           ttl 30
	        }
	        prometheus :9153
	        forward . /etc/resolv.conf {
	           max_concurrent 1000
	        }
	        cache 30
	        loop
	        reload
	        loadbalance
	    }
	kind: ConfigMap
	metadata:
	  creationTimestamp: null
	  name: coredns
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "clusterroles" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRole
	metadata:
	  creationTimestamp: null
	  name: system:coredns
	rules:
	- apiGroups:
	  - ""
	  resources:
	  - endpoints
	  - services
	  - pods
	  - namespaces
	  verbs:
	  - list
	  - watch
	- apiGroups:
	  - ""
	  resources:
	  - nodes
	  verbs:
	  - get
	- apiGroups:
	  - discovery.k8s.io
	  resources:
	  - endpointslices
	  verbs:
	  - list
	  - watch
[dryrun] Would perform action CREATE on resource "clusterrolebindings" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRoleBinding
	metadata:
	  creationTimestamp: null
	  name: system:coredns
	roleRef:
	  apiGroup: rbac.authorization.k8s.io
	  kind: ClusterRole
	  name: system:coredns
	subjects:
	- kind: ServiceAccount
	  name: coredns
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "serviceaccounts" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	kind: ServiceAccount
	metadata:
	  creationTimestamp: null
	  name: coredns
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "deployments" in API group "apps/v1"
[dryrun] Attached object:
	apiVersion: apps/v1
	kind: Deployment
	metadata:
	  creationTimestamp: null
	  labels:
	    k8s-app: kube-dns
	  name: coredns
	  namespace: kube-system
	spec:
	  replicas: 2
	  selector:
	    matchLabels:
	      k8s-app: kube-dns
	  strategy:
	    rollingUpdate:
	      maxUnavailable: 1
	    type: RollingUpdate
	  template:
	    metadata:
	      creationTimestamp: null
	      labels:
	        k8s-app: kube-dns
	    spec:
	      containers:
	      - args:
	        - -conf
	        - /etc/coredns/Corefile
	        image: k8s.gcr.io/coredns/coredns:v1.8.6
	        imagePullPolicy: IfNotPresent
	        livenessProbe:
	          failureThreshold: 5
	          httpGet:
	            path: /health
	            port: 8080
	            scheme: HTTP
	          initialDelaySeconds: 60
	          successThreshold: 1
	          timeoutSeconds: 5
	        name: coredns
	        ports:
	        - containerPort: 53
	          name: dns
	          protocol: UDP
	        - containerPort: 53
	          name: dns-tcp
	          protocol: TCP
	        - containerPort: 9153
	          name: metrics
	          protocol: TCP
	        readinessProbe:
	          httpGet:
	            path: /ready
	            port: 8181
	            scheme: HTTP
	        resources:
	          limits:
	            memory: 170Mi
	          requests:
	            cpu: 100m
	            memory: 70Mi
	        securityContext:
	          allowPrivilegeEscalation: false
	          capabilities:
	            add:
	            - NET_BIND_SERVICE
	            drop:
	            - all
	          readOnlyRootFilesystem: true
	        volumeMounts:
	        - mountPath: /etc/coredns
	          name: config-volume
	          readOnly: true
	      dnsPolicy: Default
	      nodeSelector:
	        kubernetes.io/os: linux
	      priorityClassName: system-cluster-critical
	      serviceAccountName: coredns
	      tolerations:
	      - key: CriticalAddonsOnly
	        operator: Exists
	      - effect: NoSchedule
	        key: node-role.kubernetes.io/master
	      - effect: NoSchedule
	        key: node-role.kubernetes.io/control-plane
	      volumes:
	      - configMap:
	          items:
	          - key: Corefile
	            path: Corefile
	          name: coredns
	        name: config-volume
	status: {}
[dryrun] Would perform action CREATE on resource "services" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	kind: Service
	metadata:
	  annotations:
	    prometheus.io/port: "9153"
	    prometheus.io/scrape: "true"
	  creationTimestamp: null
	  labels:
	    k8s-app: kube-dns
	    kubernetes.io/cluster-service: "true"
	    kubernetes.io/name: CoreDNS
	  name: kube-dns
	  namespace: kube-system
	  resourceVersion: "0"
	spec:
	  clusterIP: 10.96.0.10
	  ports:
	  - name: dns
	    port: 53
	    protocol: UDP
	    targetPort: 53
	  - name: dns-tcp
	    port: 53
	    protocol: TCP
	    targetPort: 53
	  - name: metrics
	    port: 9153
	    protocol: TCP
	    targetPort: 9153
	  selector:
	    k8s-app: kube-dns
	status:
	  loadBalancer: {}
[addons] Applied essential addon: CoreDNS
[dryrun] Would perform action CREATE on resource "serviceaccounts" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	kind: ServiceAccount
	metadata:
	  creationTimestamp: null
	  name: kube-proxy
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "configmaps" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	data:
	  config.conf: |-
	    apiVersion: kubeproxy.config.k8s.io/v1alpha1
	    bindAddress: 0.0.0.0
	    bindAddressHardFail: false
	    clientConnection:
	      acceptContentTypes: ""
	      burst: 0
	      contentType: ""
	      kubeconfig: /var/lib/kube-proxy/kubeconfig.conf
	      qps: 0
	    clusterCIDR: ""
	    configSyncPeriod: 0s
	    conntrack:
	      maxPerCore: null
	      min: null
	      tcpCloseWaitTimeout: null
	      tcpEstablishedTimeout: null
	    detectLocal:
	      bridgeInterface: ""
	      interfaceNamePrefix: ""
	    detectLocalMode: ""
	    enableProfiling: false
	    healthzBindAddress: ""
	    hostnameOverride: ""
	    iptables:
	      masqueradeAll: false
	      masqueradeBit: null
	      minSyncPeriod: 0s
	      syncPeriod: 0s
	    ipvs:
	      excludeCIDRs: null
	      minSyncPeriod: 0s
	      scheduler: ""
	      strictARP: false
	      syncPeriod: 0s
	      tcpFinTimeout: 0s
	      tcpTimeout: 0s
	      udpTimeout: 0s
	    kind: KubeProxyConfiguration
	    metricsBindAddress: ""
	    mode: ""
	    nodePortAddresses: null
	    oomScoreAdj: null
	    portRange: ""
	    showHiddenMetricsForVersion: ""
	    udpIdleTimeout: 0s
	    winkernel:
	      enableDSR: false
	      forwardHealthCheckVip: false
	      networkName: ""
	      rootHnsEndpointName: ""
	      sourceVip: ""
	  kubeconfig.conf: |-
	    apiVersion: v1
	    kind: Config
	    clusters:
	    - cluster:
	        certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
	        server: https://10.6.177.41:6443
	      name: default
	    contexts:
	    - context:
	        cluster: default
	        namespace: default
	        user: default
	      name: default
	    current-context: default
	    users:
	    - name: default
	      user:
	        tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
	kind: ConfigMap
	metadata:
	  annotations:
	    kubeadm.kubernetes.io/component-config.hash: sha256:1006d6c8a3ea2eec5c5b95941459a6b65853d09fb619f58de921444186e7cd85
	  creationTimestamp: null
	  labels:
	    app: kube-proxy
	  name: kube-proxy
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "daemonsets" in API group "apps/v1"
[dryrun] Attached object:
	apiVersion: apps/v1
	kind: DaemonSet
	metadata:
	  creationTimestamp: null
	  labels:
	    k8s-app: kube-proxy
	  name: kube-proxy
	  namespace: kube-system
	spec:
	  selector:
	    matchLabels:
	      k8s-app: kube-proxy
	  template:
	    metadata:
	      creationTimestamp: null
	      labels:
	        k8s-app: kube-proxy
	    spec:
	      containers:
	      - command:
	        - /usr/local/bin/kube-proxy
	        - --config=/var/lib/kube-proxy/config.conf
	        - --hostname-override=$(NODE_NAME)
	        env:
	        - name: NODE_NAME
	          valueFrom:
	            fieldRef:
	              fieldPath: spec.nodeName
	        image: k8s.gcr.io/kube-proxy:v1.24.0
	        imagePullPolicy: IfNotPresent
	        name: kube-proxy
	        resources: {}
	        securityContext:
	          privileged: true
	        volumeMounts:
	        - mountPath: /var/lib/kube-proxy
	          name: kube-proxy
	        - mountPath: /run/xtables.lock
	          name: xtables-lock
	        - mountPath: /lib/modules
	          name: lib-modules
	          readOnly: true
	      hostNetwork: true
	      nodeSelector:
	        kubernetes.io/os: linux
	      priorityClassName: system-node-critical
	      serviceAccountName: kube-proxy
	      tolerations:
	      - operator: Exists
	      volumes:
	      - configMap:
	          name: kube-proxy
	        name: kube-proxy
	      - hostPath:
	          path: /run/xtables.lock
	          type: FileOrCreate
	        name: xtables-lock
	      - hostPath:
	          path: /lib/modules
	        name: lib-modules
	  updateStrategy:
	    type: RollingUpdate
	status:
	  currentNumberScheduled: 0
	  desiredNumberScheduled: 0
	  numberMisscheduled: 0
	  numberReady: 0
[dryrun] Would perform action CREATE on resource "clusterrolebindings" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRoleBinding
	metadata:
	  creationTimestamp: null
	  name: kubeadm:node-proxier
	roleRef:
	  apiGroup: rbac.authorization.k8s.io
	  kind: ClusterRole
	  name: system:node-proxier
	subjects:
	- kind: ServiceAccount
	  name: kube-proxy
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "roles" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: Role
	metadata:
	  creationTimestamp: null
	  name: kube-proxy
	  namespace: kube-system
	rules:
	- apiGroups:
	  - ""
	  resourceNames:
	  - kube-proxy
	  resources:
	  - configmaps
	  verbs:
	  - get
[dryrun] Would perform action CREATE on resource "rolebindings" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: RoleBinding
	metadata:
	  creationTimestamp: null
	  name: kube-proxy
	  namespace: kube-system
	roleRef:
	  apiGroup: rbac.authorization.k8s.io
	  kind: Role
	  name: kube-proxy
	subjects:
	- kind: Group
	  name: system:bootstrappers:kubeadm:default-node-token
[addons] Applied essential addon: kube-proxy
```

pacoxu commented:
The behavior is quite similar to `dry-run`. So I compare the output of dry-run of `kubeadm init` with only addon phase.

```
./kubeadm init --dry-run --skip-phases=preflight,kubelet-start,control-plane,etcd,upload-certs,mark-control-plane,bootstrap-token,kubelet-finalize,wait-control-plane
```

The output is similar. However, I find that rbac resources are missing. 

```
[dryrun] Resource name: "coredns"
[dryrun] Would perform action CREATE on resource "configmaps" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	data:
	  Corefile: |
	    .:53 {
	        errors
	        health {
	           lameduck 5s
	        }
	        ready
	        kubernetes cluster.local in-addr.arpa ip6.arpa {
	           pods insecure
	           fallthrough in-addr.arpa ip6.arpa
	           ttl 30
	        }
	        prometheus :9153
	        forward . /etc/resolv.conf {
	           max_concurrent 1000
	        }
	        cache 30
	        loop
	        reload
	        loadbalance
	    }
	kind: ConfigMap
	metadata:
	  creationTimestamp: null
	  name: coredns
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "clusterroles" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRole
	metadata:
	  creationTimestamp: null
	  name: system:coredns
	rules:
	- apiGroups:
	  - ""
	  resources:
	  - endpoints
	  - services
	  - pods
	  - namespaces
	  verbs:
	  - list
	  - watch
	- apiGroups:
	  - ""
	  resources:
	  - nodes
	  verbs:
	  - get
	- apiGroups:
	  - discovery.k8s.io
	  resources:
	  - endpointslices
	  verbs:
	  - list
	  - watch
[dryrun] Would perform action CREATE on resource "clusterrolebindings" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRoleBinding
	metadata:
	  creationTimestamp: null
	  name: system:coredns
	roleRef:
	  apiGroup: rbac.authorization.k8s.io
	  kind: ClusterRole
	  name: system:coredns
	subjects:
	- kind: ServiceAccount
	  name: coredns
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "serviceaccounts" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	kind: ServiceAccount
	metadata:
	  creationTimestamp: null
	  name: coredns
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "deployments" in API group "apps/v1"
[dryrun] Attached object:
	apiVersion: apps/v1
	kind: Deployment
	metadata:
	  creationTimestamp: null
	  labels:
	    k8s-app: kube-dns
	  name: coredns
	  namespace: kube-system
	spec:
	  replicas: 2
	  selector:
	    matchLabels:
	      k8s-app: kube-dns
	  strategy:
	    rollingUpdate:
	      maxUnavailable: 1
	    type: RollingUpdate
	  template:
	    metadata:
	      creationTimestamp: null
	      labels:
	        k8s-app: kube-dns
	    spec:
	      containers:
	      - args:
	        - -conf
	        - /etc/coredns/Corefile
	        image: k8s.gcr.io/coredns/coredns:v1.8.6
	        imagePullPolicy: IfNotPresent
	        livenessProbe:
	          failureThreshold: 5
	          httpGet:
	            path: /health
	            port: 8080
	            scheme: HTTP
	          initialDelaySeconds: 60
	          successThreshold: 1
	          timeoutSeconds: 5
	        name: coredns
	        ports:
	        - containerPort: 53
	          name: dns
	          protocol: UDP
	        - containerPort: 53
	          name: dns-tcp
	          protocol: TCP
	        - containerPort: 9153
	          name: metrics
	          protocol: TCP
	        readinessProbe:
	          httpGet:
	            path: /ready
	            port: 8181
	            scheme: HTTP
	        resources:
	          limits:
	            memory: 170Mi
	          requests:
	            cpu: 100m
	            memory: 70Mi
	        securityContext:
	          allowPrivilegeEscalation: false
	          capabilities:
	            add:
	            - NET_BIND_SERVICE
	            drop:
	            - all
	          readOnlyRootFilesystem: true
	        volumeMounts:
	        - mountPath: /etc/coredns
	          name: config-volume
	          readOnly: true
	      dnsPolicy: Default
	      nodeSelector:
	        kubernetes.io/os: linux
	      priorityClassName: system-cluster-critical
	      serviceAccountName: coredns
	      tolerations:
	      - key: CriticalAddonsOnly
	        operator: Exists
	      - effect: NoSchedule
	        key: node-role.kubernetes.io/master
	      - effect: NoSchedule
	        key: node-role.kubernetes.io/control-plane
	      volumes:
	      - configMap:
	          items:
	          - key: Corefile
	            path: Corefile
	          name: coredns
	        name: config-volume
	status: {}
[dryrun] Would perform action CREATE on resource "services" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	kind: Service
	metadata:
	  annotations:
	    prometheus.io/port: "9153"
	    prometheus.io/scrape: "true"
	  creationTimestamp: null
	  labels:
	    k8s-app: kube-dns
	    kubernetes.io/cluster-service: "true"
	    kubernetes.io/name: CoreDNS
	  name: kube-dns
	  namespace: kube-system
	  resourceVersion: "0"
	spec:
	  clusterIP: 10.96.0.10
	  ports:
	  - name: dns
	    port: 53
	    protocol: UDP
	    targetPort: 53
	  - name: dns-tcp
	    port: 53
	    protocol: TCP
	    targetPort: 53
	  - name: metrics
	    port: 9153
	    protocol: TCP
	    targetPort: 9153
	  selector:
	    k8s-app: kube-dns
	status:
	  loadBalancer: {}
[addons] Applied essential addon: CoreDNS
[dryrun] Would perform action CREATE on resource "serviceaccounts" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	kind: ServiceAccount
	metadata:
	  creationTimestamp: null
	  name: kube-proxy
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "configmaps" in API group "core/v1"
[dryrun] Attached object:
	apiVersion: v1
	data:
	  config.conf: |-
	    apiVersion: kubeproxy.config.k8s.io/v1alpha1
	    bindAddress: 0.0.0.0
	    bindAddressHardFail: false
	    clientConnection:
	      acceptContentTypes: ""
	      burst: 0
	      contentType: ""
	      kubeconfig: /var/lib/kube-proxy/kubeconfig.conf
	      qps: 0
	    clusterCIDR: ""
	    configSyncPeriod: 0s
	    conntrack:
	      maxPerCore: null
	      min: null
	      tcpCloseWaitTimeout: null
	      tcpEstablishedTimeout: null
	    detectLocal:
	      bridgeInterface: ""
	      interfaceNamePrefix: ""
	    detectLocalMode: ""
	    enableProfiling: false
	    healthzBindAddress: ""
	    hostnameOverride: ""
	    iptables:
	      masqueradeAll: false
	      masqueradeBit: null
	      minSyncPeriod: 0s
	      syncPeriod: 0s
	    ipvs:
	      excludeCIDRs: null
	      minSyncPeriod: 0s
	      scheduler: ""
	      strictARP: false
	      syncPeriod: 0s
	      tcpFinTimeout: 0s
	      tcpTimeout: 0s
	      udpTimeout: 0s
	    kind: KubeProxyConfiguration
	    metricsBindAddress: ""
	    mode: ""
	    nodePortAddresses: null
	    oomScoreAdj: null
	    portRange: ""
	    showHiddenMetricsForVersion: ""
	    udpIdleTimeout: 0s
	    winkernel:
	      enableDSR: false
	      forwardHealthCheckVip: false
	      networkName: ""
	      rootHnsEndpointName: ""
	      sourceVip: ""
	  kubeconfig.conf: |-
	    apiVersion: v1
	    kind: Config
	    clusters:
	    - cluster:
	        certificate-authority: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
	        server: https://10.6.177.41:6443
	      name: default
	    contexts:
	    - context:
	        cluster: default
	        namespace: default
	        user: default
	      name: default
	    current-context: default
	    users:
	    - name: default
	      user:
	        tokenFile: /var/run/secrets/kubernetes.io/serviceaccount/token
	kind: ConfigMap
	metadata:
	  annotations:
	    kubeadm.kubernetes.io/component-config.hash: sha256:1006d6c8a3ea2eec5c5b95941459a6b65853d09fb619f58de921444186e7cd85
	  creationTimestamp: null
	  labels:
	    app: kube-proxy
	  name: kube-proxy
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "daemonsets" in API group "apps/v1"
[dryrun] Attached object:
	apiVersion: apps/v1
	kind: DaemonSet
	metadata:
	  creationTimestamp: null
	  labels:
	    k8s-app: kube-proxy
	  name: kube-proxy
	  namespace: kube-system
	spec:
	  selector:
	    matchLabels:
	      k8s-app: kube-proxy
	  template:
	    metadata:
	      creationTimestamp: null
	      labels:
	        k8s-app: kube-proxy
	    spec:
	      containers:
	      - command:
	        - /usr/local/bin/kube-proxy
	        - --config=/var/lib/kube-proxy/config.conf
	        - --hostname-override=$(NODE_NAME)
	        env:
	        - name: NODE_NAME
	          valueFrom:
	            fieldRef:
	              fieldPath: spec.nodeName
	        image: k8s.gcr.io/kube-proxy:v1.24.0
	        imagePullPolicy: IfNotPresent
	        name: kube-proxy
	        resources: {}
	        securityContext:
	          privileged: true
	        volumeMounts:
	        - mountPath: /var/lib/kube-proxy
	          name: kube-proxy
	        - mountPath: /run/xtables.lock
	          name: xtables-lock
	        - mountPath: /lib/modules
	          name: lib-modules
	          readOnly: true
	      hostNetwork: true
	      nodeSelector:
	        kubernetes.io/os: linux
	      priorityClassName: system-node-critical
	      serviceAccountName: kube-proxy
	      tolerations:
	      - operator: Exists
	      volumes:
	      - configMap:
	          name: kube-proxy
	        name: kube-proxy
	      - hostPath:
	          path: /run/xtables.lock
	          type: FileOrCreate
	        name: xtables-lock
	      - hostPath:
	          path: /lib/modules
	        name: lib-modules
	  updateStrategy:
	    type: RollingUpdate
	status:
	  currentNumberScheduled: 0
	  desiredNumberScheduled: 0
	  numberMisscheduled: 0
	  numberReady: 0
[dryrun] Would perform action CREATE on resource "clusterrolebindings" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: ClusterRoleBinding
	metadata:
	  creationTimestamp: null
	  name: kubeadm:node-proxier
	roleRef:
	  apiGroup: rbac.authorization.k8s.io
	  kind: ClusterRole
	  name: system:node-proxier
	subjects:
	- kind: ServiceAccount
	  name: kube-proxy
	  namespace: kube-system
[dryrun] Would perform action CREATE on resource "roles" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: Role
	metadata:
	  creationTimestamp: null
	  name: kube-proxy
	  namespace: kube-system
	rules:
	- apiGroups:
	  - ""
	  resourceNames:
	  - kube-proxy
	  resources:
	  - configmaps
	  verbs:
	  - get
[dryrun] Would perform action CREATE on resource "rolebindings" in API group "rbac.authorization.k8s.io/v1"
[dryrun] Attached object:
	apiVersion: rbac.authorization.k8s.io/v1
	kind: RoleBinding
	metadata:
	  creationTimestamp: null
	  name: kube-proxy
	  namespace: kube-system
	roleRef:
	  apiGroup: rbac.authorization.k8s.io
	  kind: Role
	  name: kube-proxy
	subjects:
	- kind: Group
	  name: system:bootstrappers:kubeadm:default-node-token
[addons] Applied essential addon: kube-proxy
```

wangyysde commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -46,22 +47,38 @@ const (
-	if err := CreateServiceAccount(client); err != nil {
-		return errors.Wrap(err, "error when creating kube-proxy service account")
+func EnsureProxyAddon(cfg *kubeadmapi.ClusterConfiguration, localEndpoint *kubeadmapi.APIEndpoint, client clientset.Interface, out io.Writer) error {
+	if !PrintManifestFlag {
```
Yes,I skip to print RBAC and SA . I will try to add it .

wangyysde replied:
thks.

wangyysde commented:
@neolit123 @pacoxu @brosner I have redone  this PR. Could you review it again? Thks!

pacoxu commented on cmd/kubeadm/app/cmd/phases/init/addons.go:
```diff
@@ -42,6 +45,12 @@ var (
+
+	proxyLocalFlags := pflag.NewFlagSet(options.PrintManifest, pflag.ContinueOnError)
+	proxyLocalFlags.BoolVar(&proxyaddon.PrintManifestFlag, options.PrintManifest, proxyaddon.PrintManifestFlag, "Print the addon manifests to STDOUT instead of installing them")
+
```
If we add the flag like this, the `all` will not support this flag.
```
[root@daocloud ~]# ./kubeadm init phase addon  all --print-manifest
unknown flag: --print-manifest
To see the stack trace of this error execute with --v=5 or higher
```

neolit123 replied:
I think it's ok to not support it for "all". We could do it but it introduce some minor code complexity. On the other hand users can just execute the separate addon phases with the flag.

pacoxu replied:
https://github.com/kubernetes/kubernetes/pull/109995/files#diff-872fa7d2e3a260ba4ae68c07c622fbde9f4202d8bb0d1830103c9dbf9ab3ac05R51


If so, the global variable `var PrintManifestFlag bool = false` across the packages above is easy to implement. 
/lgtm

neolit123 commented on cmd/kubeadm/app/cmd/phases/init/addons.go:
```diff
@@ -42,6 +45,12 @@ var (
 // NewAddonPhase returns the addon Cobra command
 func NewAddonPhase() workflow.Phase {
+	dnsLocalFlags := pflag.NewFlagSet(options.PrintManifest, pflag.ContinueOnError)
+	dnsLocalFlags.BoolVar(&dnsaddon.PrintManifestFlag, options.PrintManifest, dnsaddon.PrintManifestFlag, "Print the addon manifests to STDOUT instead of installing them")
```
we shouldn't use a variable from another package to populate flag values.
we can store this in a local variable in addons.go - `printManifestFlag`.
then we can pass this variable to `dnsaddon.Ensure*Addon()` functions.

wangyysde replied:
Ok, I will modify it .

neolit123 commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -47,6 +48,8 @@ const (
 	coreDNSReplicas       = 2
 )
 
+var PrintManifestFlag bool = false
```
see [1]

neolit123 commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -46,28 +48,136 @@ const (
 }
 
+//  printManifest build manifests for installing kube-proxy and output the manifest to the out
+func printManifest(cmByte []byte, dsByte []byte, out io.Writer) error {
```
i think we should add helper functions or variables that can be used for both:
- creating the objects with a client
- printing the objects

for example:
https://github.com/wangyysde/kubernetes/blob/edb2ff26a4cfbdead36bb55077b04727249d160d/cmd/kubeadm/app/phases/addons/proxy/proxy.go#L184-L189

this service account can be a variable in proxy.go called `serviceAccount`

similar all the roles and bindings from:
https://github.com/wangyysde/kubernetes/blob/edb2ff26a4cfbdead36bb55077b04727249d160d/cmd/kubeadm/app/phases/addons/proxy/proxy.go#L274-L331

can be variables.

then `printManifest()` can just marshalToYaml the objects from variables.

neolit123 replied:
if you look at the DNS printing we already have similar variables and we don't have to construct objects in two places.

wangyysde replied:
Ok, I will modify it.

wangyysde commented:
I have modified this PR. @neolit123 @pacoxu Could you review it again? Thks!

neolit123 reviewed (commented):
@wangyysde thanks for continuing to work on this.
added a few minor comments, but it seems we are going to be able to merge this soon.

neolit123 commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -47,58 +49,161 @@ const (
+		return err
 	}
 
 	fmt.Println("[addons] Applied essential addon: kube-proxy")
```
let's use `fmt.Fprintf(out, ...)` here as well.

wangyysde replied:
Ok, thks.

wangyysde commented on cmd/kubeadm/app/phases/addons/dns/dns.go:
```diff
@@ -132,6 +133,22 @@ func coreDNSAddon(cfg *kubeadmapi.ClusterConfiguration, client clientset.Interfa
+
 	if err := createCoreDNSAddon(coreDNSDeploymentBytes, coreDNSServiceBytes, coreDNSConfigMapBytes, client); err != nil {
 		return err
 	}
```
Ok .

wangyysde commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -47,58 +49,161 @@ const (
-		return errors.Wrap(err, "error when creating kube-proxy service account")
-	}
+func EnsureProxyAddon(cfg *kubeadmapi.ClusterConfiguration, localEndpoint *kubeadmapi.APIEndpoint, client clientset.Interface, out io.Writer, printManifest bool) error {
 
```
Thks.

wangyysde commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -47,58 +49,161 @@ const (
 
-// CreateServiceAccount creates the necessary serviceaccounts that kubeadm uses/might use, if they don't already exist.
-func CreateServiceAccount(client clientset.Interface) error {
+// Create Sa, RBACRules or print manifests of them to out if printManifest is true
```
Thks.

wangyysde commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -47,58 +49,161 @@ const (
-// CreateServiceAccount creates the necessary serviceaccounts that kubeadm uses/might use, if they don't already exist.
-func CreateServiceAccount(client clientset.Interface) error {
+// Create Sa, RBACRules or print manifests of them to out if printManifest is true
+func CreateSARBACRules(cmByte []byte, dsByte []byte, client clientset.Interface, out io.Writer, printManifest bool) error {
```
I have renamed the function to printOrCreateKubeProxyObjects

wangyysde commented on cmd/kubeadm/app/phases/addons/proxy/proxy.go:
```diff
@@ -47,58 +49,161 @@ const (
+		},
+	}
+
+	// Create SA, ClusterRoleBinding, Role,and RoleBinding when Print-Manifest flage not specified,
```
Thks.

wangyysde commented:
I have modified this PR. @neolit123 @pacoxu Could you review it again? Thks!

neolit123 reviewed (commented):
/lgtm
/approve

good job, thanks!

neolit123 commented:
so i tested it locally and there are some minor problems, but overall this is working as expected.
/hold

**1. the client should not be needed when we print manifests**

> kubeadm init phase addon kube-proxy --print-manifest
error execution phase addon/kube-proxy: failed to load admin kubeconfig: open /etc/kubernetes/admin.conf: no such file or directory
To see the stack trace of this error execute with --v=5 or higher

> kubeadm init phase addon coredns --print-manifest
error execution phase addon/coredns: failed to load admin kubeconfig: open /etc/kubernetes/admin.conf: no such file or directory
To see the stack trace of this error execute with --v=5 or higher

this can be avoided with a simple fix:
```diff
diff --git a/cmd/kubeadm/app/cmd/phases/init/addons.go b/cmd/kubeadm/app/cmd/phases/init/addons.go
index c31eb779fd1..47fa070ae69 100644
--- a/cmd/kubeadm/app/cmd/phases/init/addons.go
+++ b/cmd/kubeadm/app/cmd/phases/init/addons.go
@@ -90,9 +90,13 @@ func getInitData(c workflow.RunData) (*kubeadmapi.InitConfiguration, clientset.I
                return nil, nil, nil, errors.New("addon phase invoked with an invalid data struct")
        }
        cfg := data.Cfg()
-       client, err := data.Client()
-       if err != nil {
-               return nil, nil, nil, err
+       var client clientset.Interface
+       var err error
+       if !printManifest {
+               client, err = data.Client()
+               if err != nil {
+                       return nil, nil, nil, err
+               }
        }
 
        out := data.OutputWriter()
```

**2. even if we pass --print-manifest the lines:**

```
[addons] Applied essential addon: kube-proxy / coredns
```

are printed. we should only print them if we are actually applying the addons.

**3. when the manifests are printed there are extra new lines around the `---`. can we avoid these somehow?**

e.g. currently it's:
```
...
  verbs:
  - get

---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  creationTimestamp: null
  name: kube-proxy
  namespace: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: kube-proxy
subjects:
- kind: Group
  name: system:bootstrappers:kubeadm:default-node-token

---

kind: ConfigMap
apiVersion: v1
metadata:
  name: kube-proxy
```

should be:

```
...
  verbs:
  - get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  creationTimestamp: null
  name: kube-proxy
  namespace: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: kube-proxy
subjects:
- kind: Group
  name: system:bootstrappers:kubeadm:default-node-token
---
kind: ConfigMap
apiVersion: v1
metadata:
  name: kube-proxy
```

wangyysde commented:
/test pull-kubernetes-e2e-kind

wangyysde commented:
@neolit123 @pacoxu 
I have modified this PR. Could you review it again? Thanks.

> 1. the client should not be needed when we print manifests

Thks. I have modified it.

> 2. even if we pass --print-manifest the lines:

This bug has be fixed. thks.

> when the manifests are printed there are extra new lines around the ---. can we avoid these somehow?

New lines have been omitted.

neolit123 commented:
/retest
